File: C-library/examples/Mandelbrot/Mandelbrot.r
#! /usr/bin/env rebol
REBOL [
	Title:		"Mandelbrot fractal ASCII renderer"
	Author:		"Kaj de Vos"
	Rights:		"Copyright (c) 2012 Kaj de Vos"
	License: {
		PD/CC0
		http://creativecommons.org/publicdomain/zero/1.0/
	}
;	Needs:		"REBOL 3 or 2"
	Tabs:		4
]

threshold: 16.0
max-iterations: 50'000

mandelbrot: func [
	x		[decimal!]
	y		[decimal!]
	/local	i cr ci zr zi zr2 zi2 value
][
	cr: y - 0.5
	ci: x
	zr: zi: 0.0

	repeat i max-iterations [
		value: zr * zi
		zr: (zr2: zr * zr) - (zi2: zi * zi) + cr
		zi: value + value + ci

		if zr2 + zi2 > threshold [return i]
	]
	0
]

start: now/time/precise

for y -39 39 1 [
	for x -39 39 1 [
		prin either zero? mandelbrot x / 40.0  y / 40.0 [#"*"] [#" "]
	]
	print ""
]

print ["Elapsed time:" now/time/precise - start]

File: C-library/examples/Mandelbrot/Mandelbrot.c
/*
	Mandelbrot fractal ASCII renderer
	Copyright (c) 2012 Kaj de Vos
	Adapted from Erik Wrenholt
		http://www.timestretch.com/article/mandelbrot_fractal_benchmark
	License: PD/CC0
		http://creativecommons.org/publicdomain/zero/1.0/
*/

#include <stdio.h>
#include <time.h>

#define threshold		16.0
#define max_iterations	50000

int mandelbrot (
	double x,
	double y
){
	double cr = y - 0.5;
	double ci = x;
	double zr = 0.0, zi = 0.0;
	int i = 1;

	do {
		double value = zr * zi;
		double zr2 = zr * zr;
		double zi2 = zi * zi;
		zr = zr2 - zi2 + cr;
		zi = value + value + ci;

		if (zr2 + zi2 > threshold) return i;
	} while (++i <= max_iterations);

	return 0;
}

int main (int argc, const char * argv []) {
	time_t start = time (NULL);

	for (int y = -39; y <= 39; y++) {
		for (int x = -39; x <= 39; x++) {
			putchar (mandelbrot (x / 40.0, y / 40.0) == 0 ? '*' : ' ');
		}
		putchar ('\n');
	}

	printf ("Elapsed time: %f\nProcessor time: %f\n",
		difftime (time (NULL), start),
		(double) clock () / CLOCKS_PER_SEC
	);
}
